docs/proposals/001-in-place-upgrades.md

```diff
@@ -0,0 +1,221 @@
+
+However at the time of writing CAPI does not have support for changing upgrade strategies which means changes to the `version` fields trigger a rolling update.
+
+This behaviour can be adjusted on `ControlPlane` objects as our provider has more/full control but can not be easily adjusted on `MachineDeployment` objects which causes issues for worker nodes.
+
+Switching to using the `version` field should take place when upstream implements support for different upgrade strategies.
+
+## Out of scope
+<!--
+This section MUST reference any work that is out of scope for this proposal.
+Out of scope items are typically unknowns that we do not yet have a clear idea
+of how to solve, so we explicitly do not tackle them until we have more
+information.
+
+This section is very useful to help guide the implementation details section
+below, or serve as reference for future proposals.
+-->
+
+none
+
+# Implementation Details
+
+## API Changes
+<!--
+This section MUST mention any changes to the k8sd API, or any additional API
+endpoints (and messages) that are required for this proposal.
+
+Unless there is a particularly strong reason, it is preferable to add new v2/v3
+APIs endpoints instead of breaking the existing APIs, such that API clients are
+not affected.
+-->
+### `POST /x/capi/snap-refresh`
+
+```go
+type SnapRefreshRequest struct {
+	// Channel is the channel to refresh the snap to.
+	Channel string
+	// Revision is the revision number to refresh the snap to.
+	Revision string
+	// LocalPath is the local path to use to refresh the snap.
+	LocalPath string
+}
+```
+
+`POST /x/capi/snap-refresh` performs the in-place upgrade with the given options. The upgrade can be either done with a `Channel`, `Revision` or a local `*.snap` file provided via `LocalPath`.
+
+This endpoint should use `ValidateCAPIAuthTokenAccessHandler("capi-auth-token")` for authentication.
+
+## Bootstrap Provider Changes
+<!--
+This section MUST mention any changes to the bootstrap provider.
+-->
+
+A machine controller called `MachineReconciler` is added which would perform the in-place upgrade if `k8sd.io/in-place-upgrade` annotation is set on the machine.
+
+The controller would use the value of this annotation to make an endpoint call to the `/x/capi/snap-refresh` through `k8sd-proxy`. 
+
+The result of this operation will be communicated back to the user via the `k8sd.io/in-place-upgrade-status` annotation. Values being:
+
+* `done` for a successful upgrade
+* `failed` for a failed upgrade
+
+A failed upgrade could be re-triggered by removing the `k8sd.io/in-place-upgrade` annotation and re-adding it to the machine.
+
+The `ck8sconfig_controller` should check for the `k8sd.io/in-place-upgrade` annotation both on the `Machine` and on the owner `Cluster` object. The value of one of these annotations should be used instead of the `version` field while generating a cloud-init script for a new machine. The annotation on the `Machine` object should take precedence.
```

Comment:
I wonder if by default an update on the cluster annotation should update all nodes. Each machine would always have the version as an annotation or something like `no-update`.

Then we could retry updating the machine version to the cluster-version until they are the same?

I think this could be useful in a sense that the control over the update is closer to the management cluster level rather than the individual level.

Reply:
Let's not do it. Separation of concerns - you don't want to have upgrade logic spread over multiple controllers.

Reply:
Agreeing with @bschimke95 here, the upgrades are scoped to `Machine`'s. The `Cluster` level annotation is simply a helper to offload the work of annotating all cluster machines manually.